### PR TITLE
refactor: L2 constructs should not use the "Token" type (awslint:props-no-tokens)

### LINKED ISSUE #2434: awslint:props-no-tokens
Props should not use the "Token" type

### PR DESCRIPTION
**Pull request by shivlaks**
Adds a new awslint:props-no-tokens rule which validates that props do not use the "Token" type.

This is in accordance with the new AWS Construct Library guidelines.

**Fixes #2434**

----

### Pull Request Checklist

* [x] Testing
  - Unit test added (prefer not to modify an existing test, otherwise, it's probably a breaking change)
  - __CLI change?:__ coordinate update of integration tests with team
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [x] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
  - __Design__: For significant features, design document added to `design` folder
* [x] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** and module name in parens, which will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [x] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

File: tools/awslint/lib/rules/construct.ts [hunk 1]
```diff
@@ -227,4 +227,29 @@ constructLinter.add({
...
+  code: 'props-no-tokens',
+  message: 'props should not use the "Token" type',
+  eval: e => {
+    if (!e.ctx.propsType) { return; }
+    if (!e.ctx.hasPropsArgument) { return; }
+
+    // this rule only applies to L2 constructs
+    if (e.ctx.classType.name.startsWith('Cfn')) { return; }
+
+    for (const property of e.ctx.propsType.ownProperties) {
```
Comment (eladb) [inline tools/awslint/lib/rules/construct.ts]: why `ownProperties` and not `allProperties`?

File: tools/awslint/lib/rules/construct.ts [hunk 2]
```diff
@@ -227,4 +227,29 @@ constructLinter.add({
...
+      const typeRef = property.type;
+      let fqn = typeRef.fqn;
+
+      if (typeRef.arrayOfType) {
+        fqn = typeRef.arrayOfType.fqn;
+      } else if (typeRef.mapOfType) {
+        fqn = typeRef.mapOfType.fqn;
+      }
+
+      e.assert(!(fqn === '@aws-cdk/cdk.Token'), `${e.ctx.propsFqn}.${property.name}`);
```
Comment (eladb) [inline tools/awslint/lib/rules/construct.ts]: We should probably put all these FQNs into a single file in awslint so if we modify them we have a good place to verify. I would also argue that we should look up this FQN in the typesystem to ensure that it actually exists.

File: tools/awslint/lib/rules/common.ts [hunk 3]
```diff
@@ -1 +1,5 @@
 export const CORE_MODULE = '@aws-cdk/cdk';
+
+export const enum TypeFqn {
+    Token = '@aws-cdk/cdk.Token'
```
Comment (eladb) [inline tools/awslint/lib/rules/common.ts]: Do you mind moving all FQNs to here? they are spread across awslint...

File: tools/awslint/lib/rules/construct.ts [hunk 4]
```diff
@@ -249,7 +249,10 @@ constructLinter.add({
         fqn = typeRef.mapOfType.fqn;
       }
 
-      e.assert(!(fqn === '@aws-cdk/cdk.Token'), `${e.ctx.propsFqn}.${property.name}`);
```
Comment (eladb) [inline tools/awslint/lib/rules/construct.ts]: Not sure that's what I would expect... Say we rename cdk.Token to something else. Then basically this linter rule will completely be ignored. I was looking for a way to be defensive around FQN changes given this is very loosely coupled. 


This is a general problem we have with all of the core types we refer to in awslint, so I am wondering if we can create a shared facility that will validate that the core types we want to refer to actually exist in the typesystem:

```ts
// pseudo code:

export class CoreTypes {
  constructor(typesystem: reflect.TypeSystem) {
    if (!typesystem.includesAssembly('@aws-cdk/cdk')) { disable-all-checks })
    for each fqn of CORE_TYPES_FQN {
      if (!typesystem.tryFindFqn(fqn)) {
         throw new Error('cannot find core FQN type ${fqn}');
     }
   }
  }

  public get tokenClass() { return this.typesystem.findClass(CORE_TYPES_FQN.Token); }
  //...
}
```

File: tools/awslint/lib/rules/construct.ts (hunk 1 above)
Comment (shivlaks) [reply tools/awslint/lib/rules/construct.ts]: good catch - updated.

File: tools/awslint/lib/rules/common.ts (hunk 3 above)
Comment (shivlaks) [reply tools/awslint/lib/rules/common.ts]: no problem - will pick them all up

File: tools/awslint/lib/rules/construct.ts [hunk 5]
```diff
@@ -227,4 +207,32 @@ constructLinter.add({
...
+
+      if (typeRef.arrayOfType) {
+        fqn = typeRef.arrayOfType.fqn;
+      } else if (typeRef.mapOfType) {
+        fqn = typeRef.mapOfType.fqn;
+      }
+
+      const found = (fqn && e.ctx.sys.tryFindFqn(fqn));
+      if(found) {
+        e.assert(!(fqn === e.ctx.core.tokenClass.fqn), `${e.ctx.propsFqn}.${property.name}`);
```
Comment (rix0rrr) [inline tools/awslint/lib/rules/construct.ts]: The assert message should probably include a descriptive text in addition to just the property name.

File: tools/awslint/lib/rules/core-types.ts [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+  private readonly sys: TypeSystem;
+
+  constructor(sys: reflect.TypeSystem) {
+    this.sys = sys;
+    if (!sys.includesAssembly(CORE_MODULE)) {
+      // disable-all-checks
+      return;
+    }
+
+    for (const fqn of Object.values(CoreTypesFqn)) {
```
Comment (rix0rrr) [inline tools/awslint/lib/rules/core-types.ts]: I honestly think we don't need this validation.

File: tools/awslint/lib/rules/construct.ts [hunk 7]
```diff
@@ -227,4 +207,32 @@ constructLinter.add({
...
+
+constructLinter.add({
+  code: 'props-no-tokens',
+  message: 'props should not use the "Token" type',
+  eval: e => {
+    if (!e.ctx.propsType) { return; }
+    if (!e.ctx.hasPropsArgument) { return; }
+
+    // this rule only applies to L2 constructs
+    if (e.ctx.classType.name.startsWith('Cfn')) { return; }
```
Comment (rix0rrr) [inline tools/awslint/lib/rules/construct.ts]: `IsCfnResource`?

Comment (shivlaks) [reply tools/awslint/lib/rules/construct.ts]: good catch. Done.

File: tools/awslint/lib/rules/construct.ts (hunk 5 above)
Comment (shivlaks) [reply tools/awslint/lib/rules/construct.ts]: The property name is just the scope of the rule. The text describing the violation comes from the message field. 

eg:

```
error: [awslint:props-no-tokens:@aws-cdk/aws-resource.ResourceProps.badParameter] props should not use the "Token" type 
```

Am I missing something?

Comment (rix0rrr) [reply tools/awslint/lib/rules/construct.ts]: Nope, you are correct, just misunderstood the purpose of the argument.

**Pull request merged**